utils/server_auth.go

```diff
@@ -0,0 +1,342 @@
+		}
+
+		for _, tokenScope := range strings.Split(scope, " ") {
+			// As long as there's one scope in the token that matches the pool of desriedScopes
+			// we say it's valid
+			if scopeContains(tokenScope, anyScopes) {
+				return nil
+			}
+		}
+		return jwt.NewValidationError(errors.New(fmt.Sprint("Token does not contain any of the scopes: ", anyScopes)))
+	})
+}
+
+// Checks that the given token was signed by the federation jwk and also checks that the token has the expected scope
+func (a AuthCheckImpl) FederationCheck(c *gin.Context, strToken string, anyOfTheScopes []string) error {
+	var bKey *jwk.Key
+
+	fedURL := param.Federation_DiscoveryUrl.GetString()
+	token, err := jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+
+	if err != nil {
+		return err
+	}
+
+	if fedURL != token.Issuer() {
+		return errors.New(fmt.Sprint("Issuer is not a federation: ", token.Issuer()))
+	}
+
+	fedURIFile := param.Federation_JwkUrl.GetString()
+	ctx := context.Background()
+	if federationJWK == nil {
+		client := &http.Client{Transport: config.GetTransport()}
+		federationJWK = jwk.NewCache(ctx)
+		if err := federationJWK.Register(fedURIFile, jwk.WithRefreshInterval(15*time.Minute), jwk.WithHTTPClient(client)); err != nil {
+			return errors.Wrap(err, "Failed to register cache for federation's public JWKS")
+		}
+	}
+
+	jwks, err := federationJWK.Get(ctx, fedURIFile)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get federation's public JWKS")
+	}
+	key, ok := jwks.Key(0)
+	if !ok {
+		return errors.Wrap(err, "Failed to get the first key of federation's public JWKS")
+	}
+	bKey = &key
+	var raw ecdsa.PrivateKey
+	if err = (*bKey).Raw(&raw); err != nil {
+		return errors.Wrap(err, "Failed to get raw key of the federation JWK")
+	}
+
+	parsed, err := jwt.Parse([]byte(strToken), jwt.WithKey(jwa.ES256, raw.PublicKey))
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to verify JWT by federation's key")
+	}
+
+	scopeValidator := createScopeValidator(anyOfTheScopes)
+	if err = jwt.Validate(parsed, jwt.WithValidator(scopeValidator)); err != nil {
+		return errors.Wrap(err, "Failed to verify the scope of the token")
+	}
+
+	c.Set("User", "Federation")
+	return nil
+}
+
+// Checks that the given token was signed by the issuer jwk (the one from the server itself) and also checks that
+// the token has the expected scope
+func (a AuthCheckImpl) IssuerCheck(c *gin.Context, strToken string, anyOfTheScopes []string) error {
+	token, err := jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+	if err != nil {
+		return errors.Wrap(err, "Invalid JWT")
+	}
+
+	serverURL := param.Server_ExternalWebUrl.GetString()
+	if serverURL != token.Issuer() {
+		if param.Origin_Url.GetString() == token.Issuer() {
+			return errors.New(fmt.Sprint("Wrong issuer; expect the issuer to be the server's web address but got Origin.URL, " + token.Issuer()))
+		} else {
+			return errors.New(fmt.Sprint("Issuer is not server itself: ", token.Issuer()))
```

Comment:
This isn't something I know how to grapple with at the moment, but in general we don't require that the issuer is the same as the server. We do want to allow people to use their own issuers instead of the built-in issuer we provide. I don't know if there's a change for this, but just something to think about moving forward because eventually we'll have to come back to these sections.

Reply:
Will add comment as a hint for the future